src/lib/users/ssh_authorized_keyring.rb

```diff
@@ -0,0 +1,213 @@
+        authorized_keys = SSHAuthorizedKeysFile.new(path).keys
+        keys[home] = authorized_keys unless authorized_keys.empty?
+        log.info "Read #{authorized_keys.size} keys from #{path}"
+        !authorized_keys.empty?
+      end
+
+      # Write user keys to the given file
+      #
+      # If SSH_DIR does not exist in the given directory, it will be
+      # created inheriting owner/group and setting permissions to SSH_DIR_PERM.
+      #
+      # @param path [String] User's home directory
+      # @return [Boolean] +true+ if keys were written; +false+ otherwise
+      def write_keys(home)
+        return false if keys[home].nil?
+        if !FileUtils::Exists(home)
+          log.error("Home directory '#{home}' does not exist!")
+          raise HomeDoesNotExist
+        end
+        owner, group = perms_from(home)
+        if create_ssh_dir(home, owner, group, SSH_DIR_PERMS)
+          write_file(home, owner, group, AUTHORIZED_KEYS_PERMS)
+        else
+          log.error("SSH directory does not exist and could not be created: giving up")
+          return false
+        end
+      end
+
+      private
+
+      # @return [String] Relative path to the SSH directory inside users' home
+      SSH_DIR = ".ssh".freeze
+      # @return [String] Authorized keys file name
+      AUTHORIZED_KEYS_FILE = "authorized_keys".freeze
+      # @return [String] Permissions to be set on SSH_DIR directory
+      SSH_DIR_PERMS = "0700".freeze
+      # @return [String] Permissions to be set on `authorized_keys` file
+      AUTHORIZED_KEYS_PERMS = "0600".freeze
+
+      # Determine the path to the user's SSH directory
+      #
+      # @param home [String] Home directory
+      # @return [String] Path to the user's SSH directory
+      #
+      # @see SSH_DIR
+      def ssh_dir_path(home)
+        File.join(home, SSH_DIR)
+      end
+
+      # Determine the path to the user's authorized keys file
+      #
+      # @param home [String] Home directory
+      # @return [String] Path to authorized keys file
+      #
+      # @see SSH_DIR
+      # @see AUTHORIZED_KEYS_FILE
+      #
+      # @see #ssh_dir_path
+      def authorized_keys_path(home)
+        File.join(ssh_dir_path(home), AUTHORIZED_KEYS_FILE)
+      end
+
+      # Find or creates the SSH directory
+      #
+      # This method sets up the SSH directory (usually .ssh). Although only 1
+      # level is needed (as SSH directory lives under $HOME/.ssh), this code
+      # should support changing SSH_DIR to something like `.config/ssh`.
+      #
+      # @param home  [String] Home directory where SSH directory must be created
+      # @param owner [Fixnum] Owner's UID
+      # @param group [Fixnum] Owner's GID
+      # @param perms [String] Permissions (in form "0700")
+      # @return [String] Returns the path to the first created directory
+      def create_ssh_dir(home, owner, group, perms)
+        ssh_dir = ssh_dir_path(home)
+        return ssh_dir if FileUtils::Exists(ssh_dir)
```

Comment:
I'm thinking about the security issues in the code. When we are writing into the user's home we should consider it as a not trusted "input". E.g. users could make `.ssh` a symlink pointing to into another location possibly rewriting some files.

You could easily point to another user's home and override that SSH key and gain access to that account! :boom:  That's not a problem in AutoInstallation, but could be a problem in the AutoUpgrade mode...

I think here we should simply check for a symlink and abort (raise exception) in that case.

Additionaly it would be probably good to call `set_owner_and_perms(ssh_dir,...)` always to make sure the directory has correct permissions. (Or check them and abort if they are wrong.)